### PR TITLE
backport: merge bitcoin#20146, #20477, #20624, #19972, #20724, #19884, #21165, #20721, #21254, partial bitcoin#19829 (networking backports)

### PR DESCRIPTION
**Pull request by kwvg**
## Additional Information

* [bitcoin#19884](https://github.com/bitcoin/bitcoin/pull/19884) doesn't seem to play nice on its own and necessitated two more backports (namely [bitcoin#21165](https://github.com/bitcoin/bitcoin/pull/21165) and [bitcoin#21254](https://github.com/bitcoin/bitcoin/pull/21254)) before it did.
  * Trying to find why that is has been time consuming but there doesn't seem to be a concrete answer. If running the daemon normally (`dashd --regtest --dnsseed=1`), the functionality behaves as expected but the test still fails.

  * The closest explanation is that our OOO backports with relation to mocking time could explain why it isn't working as expected due to debug statements I added that always shown the time delta between each "enough time has passed" check was 0 seconds even when the log was advancing forward in time.

* The usage of `dnsseed=0` stems from [bitcoin#16551](https://github.com/bitcoin/bitcoin/pull/16551) (link to diff in commit comment), a backport that was skipped due to complexity. Though, some aspects of the PR have made it with [dash#3946](https://github.com/dashpay/dash/pull/3946).

* [bitcoin#19829](https://github.com/bitcoin/bitcoin/pull/19829) does away with `CConnman::ForEachNode` usage in `PeerManagerImpl::UpdatedBlockTip` ([source](https://github.com/bitcoin/bitcoin/pull/19829/files#diff-6875de769e90cec84d2e8a9c1b962cdbcda44d870d42e4215827e599e11e90e3R1318-R1328)). Dash cannot do the same and continues to use `ForEachNode` as we use the `CNode::CanRelay` check, which is not accessible through the `Peer` struct.

    * It would be valuable to find values that are Dash-specific (like `m_masternode_connection`) and migrate them to `Peer` to avoid Dash-specific patches and closer alignment with upstream.

## Breaking Changes

Potential change in behaviour in the GUI and RPC. 

In RPC, `getpeerinfo` will now display `pingwait` and `startingheight` if `fStateStats` is true (earlier behaviour was unconditional). `startingheight` has been placed below `banscore` (earlier behaviour placed it above). In the GUI (Qt), `peerHeight` and `peerPingWait` are subject to similar conditionality as mentioned earlier.

No changes in protocol or consensus. Changes are primarily related to refactoring, cleaning up, improving networking code and adding a new flag (`-fixedseeds`).

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated relevant unit/integration/functional/e2e tests
- [x] I have made corresponding changes to the documentation **(note: N/A)**
- [x] I have assigned this pull request to a milestone _(for repository code-owners and collaborators only)_

### CODE REVIEW

File: src/net_processing.cpp [hunk 1]
```diff
@@ -1786,13 +1817,15 @@ void PeerManagerImpl::UpdatedBlockTip(const CBlockIndex *pindexNew, const CBlock
     }
 
     // Relay to all peers
-    m_connman.ForEachNode([&vHashes](CNode* pnode) {
-        if (!pnode->CanRelay()) {
-            return;
-        }
-        LOCK(pnode->cs_inventory);
+    m_connman.ForEachNode([this, &vHashes](CNode* pnode) {
+        if (!pnode->CanRelay()) return;
```
Comment (UdjinM6) [inline src/net_processing.cpp:1847]: 19829: add TODO and/or mark it `partial`?

File: src/net.cpp [hunk 2]
```diff
@@ -1122,6 +1121,7 @@ bool CConnman::AttemptToEvictConnection()
     LOCK(cs_vNodes);
     for (CNode* pnode : vNodes) {
         if (pnode->GetId() == *node_id_to_evict) {
+            LogPrint(BCLog::NET, "selected %s connection for eviction peer=%d; disconnecting\n", pnode->ConnectionTypeAsString(), pnode->GetId());
```
Comment (UdjinM6) [inline src/net.cpp:1111]: 20724: should probably use `NET_NETCONN` here like we do in similar cases to be able to log these with `netconn` category which helps getting connection specific output only without all the `net` spam.

File: src/net.cpp [hunk 3]
```diff
@@ -1195,7 +1195,7 @@ void CConnman::CreateNodeFromAcceptedSocket(SOCKET hSocket,
     }
 
     if (!fNetworkActive) {
-        LogPrintf("%s: not accepting new connections\n", strDropped);
+        LogPrint(BCLog::NET, "%s: not accepting new connections\n", strDropped);
```
Comment (UdjinM6) [inline src/net.cpp:1185]: 20724: same

File: src/net.cpp [hunk 4]
```diff
@@ -3814,6 +3814,7 @@ bool CConnman::DisconnectNode(const std::string& strNode)
 {
     LOCK(cs_vNodes);
     if (CNode* pnode = FindNode(strNode)) {
+        LogPrint(BCLog::NET, "disconnect by address%s matched peer=%d; disconnecting\n", (fLogIPs ? strprintf("=%s", strNode) : ""), pnode->GetId());
```
Comment (UdjinM6) [inline src/net.cpp:3817]: 20724: same

File: src/net.cpp [hunk 5]
```diff
@@ -3826,6 +3827,7 @@ bool CConnman::DisconnectNode(const CSubNet& subnet)
     LOCK(cs_vNodes);
     for (CNode* pnode : vNodes) {
         if (subnet.Match(pnode->addr)) {
+            LogPrint(BCLog::NET, "disconnect by subnet%s matched peer=%d; disconnecting\n", (fLogIPs ? strprintf("=%s", subnet.ToString()) : ""), pnode->GetId());
```
Comment (UdjinM6) [inline src/net.cpp:3830]: 20724: same

File: src/net.cpp [hunk 6]
```diff
@@ -3843,6 +3845,7 @@ bool CConnman::DisconnectNode(NodeId id)
     LOCK(cs_vNodes);
     for(CNode* pnode : vNodes) {
         if (id == pnode->GetId()) {
+            LogPrint(BCLog::NET, "disconnect by id peer=%d; disconnecting\n", pnode->GetId());
```
Comment (UdjinM6) [inline src/net.cpp:3848]: 20724: same

File: src/net_processing.cpp [hunk 7]
```diff
@@ -3143,6 +3144,7 @@ void PeerManagerImpl::ProcessMessage(
 
         // Feeler connections exist only to verify if address is online.
         if (pfrom.IsFeelerConn()) {
+            LogPrint(BCLog::NET, "feeler connection completed peer=%d; disconnecting\n", pfrom.GetId());
```
Comment (UdjinM6) [inline src/net_processing.cpp:3178]: 20724: same

File: src/net_processing.cpp [hunk 8]
```diff
@@ -3211,6 +3216,7 @@ void PeerManagerImpl::ProcessMessage(
         if (pfrom.fSuccessfullyConnected) {
             // Disconnect peers that send SENDADDRV2 message after VERACK; this
             // must be negotiated between VERSION and VERACK.
+            LogPrint(BCLog::NET, "sendaddrv2 received after verack from peer=%d; disconnecting\n", pfrom.GetId());
```
Comment (UdjinM6) [inline src/net_processing.cpp:3250]: 20724: same

File: src/net_processing.cpp [hunk 9]
```diff
@@ -3299,8 +3306,10 @@ void PeerManagerImpl::ProcessMessage(
         m_addrman.Add(vAddrOk, pfrom.addr, 2 * 60 * 60);
         if (vAddr.size() < 1000)
             pfrom.fGetAddr = false;
-        if (pfrom.IsAddrFetchConn())
+        if (pfrom.IsAddrFetchConn()) {
+            LogPrint(BCLog::NET, "addrfetch connection completed peer=%d; disconnecting\n", pfrom.GetId());
```
Comment (UdjinM6) [inline src/net_processing.cpp:3341]: 20724: same

File: src/net_processing.cpp [hunk 10]
```diff
@@ -4318,6 +4327,7 @@ void PeerManagerImpl::ProcessMessage(
 
     if (msg_type == NetMsgType::FILTERLOAD) {
         if (!(pfrom.GetLocalServices() & NODE_BLOOM)) {
+            LogPrint(BCLog::NET, "filterload received despite not offering bloom services from peer=%d; disconnecting\n", pfrom.GetId());
```
Comment (UdjinM6) [inline src/net_processing.cpp:4360]: 20724: same

File: src/net_processing.cpp [hunk 11]
```diff
@@ -4340,6 +4350,7 @@ void PeerManagerImpl::ProcessMessage(
 
     if (msg_type == NetMsgType::FILTERADD) {
         if (!(pfrom.GetLocalServices() & NODE_BLOOM)) {
+            LogPrint(BCLog::NET, "filteradd received despite not offering bloom services from peer=%d; disconnecting\n", pfrom.GetId());
```
Comment (UdjinM6) [inline src/net_processing.cpp:4383]: 20724: same

File: src/net_processing.cpp [hunk 12]
```diff
@@ -4367,6 +4378,7 @@ void PeerManagerImpl::ProcessMessage(
 
     if (msg_type == NetMsgType::FILTERCLEAR) {
         if (!(pfrom.GetLocalServices() & NODE_BLOOM)) {
+            LogPrint(BCLog::NET, "filterclear received despite not offering bloom services from peer=%d; disconnecting\n", pfrom.GetId());
```
Comment (UdjinM6) [inline src/net_processing.cpp:4411]: 20724: same

File: src/net.cpp (hunk 2 above)
Comment (kwvg) [reply src/net.cpp]: Resolved in latest push

File: src/net_processing.cpp (hunk 1 above)
Comment (kwvg) [reply src/net_processing.cpp:1848]: Resolved in latest push

File: src/net.cpp [hunk 13]
```diff
@@ -930,21 +930,6 @@ size_t CConnman::SocketSendData(CNode& node)
     return nSentSize;
 }
 
-struct NodeEvictionCandidate
```
Comment (PastaPastaPasta) [inline src/net.cpp:934]: 20477: missing added optional include? probably compiles fine due to other headers already including; but we should IWYU

File: src/net.cpp [hunk 14]
```diff
@@ -1126,10 +1056,72 @@ bool CConnman::AttemptToEvictConnection()
...
+ *   for each of several distinct characteristics which are difficult
+ *   to forge.  In order to partition a node the attacker must be
+ *   simultaneously better at all of them than honest peers.
+ */
+bool CConnman::AttemptToEvictConnection()
+{
+    std::vector<NodeEvictionCandidate> vEvictionCandidates;
+    {
+        LOCK(cs_vNodes);
+
```
Comment (PastaPastaPasta) [inline src/net.cpp:1061]: 20477: nit whitespace diff with upstream could result in conflicts in future

**Review by PastaPastaPasta — COMMENTED** [review]
Mostly LGTM

**Review by PastaPastaPasta — APPROVED** [review]
utACK 99e87c98fab273965b05ea2b2d366bf925bca234

**Comment by kwvg** [comment]
Branch re-pushed to have commits arranged in order of upstream merger. No changes in diff or conflicts caused by re-org.

**Review by PastaPastaPasta — APPROVED** [review]
re-utACK 0d9046586baa035d8568aa6368f307b18a58b678

**Review by UdjinM6 — APPROVED** [review]
utACK

**Pull request merged**